### DISCUSSION TITLE
Adding directives to schema without being directly consumed by a type

### DISCUSSION THREAD

**Discussion by Barry-Zou** [ExpediaGroup/graphql-kotlin#1455] [Q&A] [answered]
Hi I saw in the documentation [here](https://opensource.expediagroup.com/graphql-kotlin/docs/schema-generator/customizing-schemas/advanced-features) that we are able to add additional custom types to the schema without being directly consumed by any existing types. 

My question is there a way to do the same for directives? E.g. I have some directives that are currently not being referenced by any types in the schema but I would like to include them in the outputted GraphQLSchema. (Apologies if this seems something obvious that I missed in the docs)

Thanks!

**Comment by dariuszkuc** [comment, marked as answer]
Hello 👋 
It is doable but will require some manual code.

You could either 
* extend `SchemaGenerator/FederatedSchemaGenerator` and transform your schema in `generateSchema`, i.e.

```kotlin
fun generateSchema(...) {
  val originalSchema = super.generateSchema(...)
  val updatedSchema = originalSchema.transform { builder ->
     // your logic goes here
  }
}
```

* provide custom [SchemaGeneratorHooks](https://opensource.expediagroup.com/graphql-kotlin/docs/schema-generator/customizing-schemas/generator-config#schemageneratorhooks) (or `FederatedSchemaGeneratorHooks`) and add your directives in the `willBuildSchema(builder: builder: GraphQLSchema.Builder)` hook

In either case you would need to manually create your directive definitions and them to the builder.

```kotlin
builder.additionalDirective(<your directive definition goes here>)
```
